### PR TITLE
SimpleParser NOT and other operations

### PR DESCRIPTION
**Pull request by nelio2k**
In order to get rid of my merge commits, I had to do a rebase and cherry-pick and with them, a few  conflicts.

- Found and fixed an issue where false JSON values were perceived as true.
- SimpleParser added a NOT operator, which is used in "!=", "<=", and ">=" operations. Utilizing Brett's code changes.
- Discussed with Brett about changing GreaterThanEq in matcher and associates to just Greater... since <= and >= can be accomplished with inverse of "<" and ">".
- Beefed up simpleParser unit test to test all the changes mentioned above. The scope of the simpleParser unit test should really be testing just the simpleParser, but it's slowly overstepping its boundary into matcher testing. Not a big deal for now but may address it later.

Note that this PR supersedes couchbaselabs/gojsonsm#13

### CODE REVIEW

File: matcher.go [hunk 1]
```diff
@@ -366,6 +366,7 @@ func (m *Matcher) Match(data []byte) (bool, error) {
 		return false, err
 	}
 
+	//	fmt.Printf("Parse node: %v\n", m.def.ParseNode)
```
Comment (brett19) [inline matcher.go]: This is not part of the actual code, I don't think its intended to be included.

File: expression_json.go [hunk 2]
```diff
@@ -243,6 +243,8 @@ func parseJsonSubexpr(data []interface{}) (Expression, error) {
 		return parseJsonLessThan(data)
 	case "greaterequal":
 		return parseJsonGreaterEquals(data)
+	case "not":
```
Comment (brett19) [inline expression_json.go]: This is a duplicate case statement.  Not is handled above.

File: transform.go [hunk 3]
```diff
@@ -48,15 +48,16 @@ func (expr mergeExpr) RootRefs() []FieldExpr {
...
-	VarIdx          VariableID
-	BucketIdx       BucketID
-	RootExec        *ExecNode
-	RootTree        binTree
-	NodeMap         map[int]*ExecNode
-	ActiveExec      *ExecNode
-	ActiveBucketIdx BucketID
-	MaxDepth        int
-	CurDepth        int
+	VarIdx              VariableID
```
Comment (brett19) [inline transform.go]: These lines don't appear to actually be related to this change.

File: transform.go [hunk 4]
```diff
@@ -79,19 +80,23 @@ func opTypeToString(value OpType) string {
...
 		return "in"
 	}
-
 	return "??unknown??"
 }
 
 type OpNode struct {
 	BucketIdx BucketID
 	Op        OpType
+	OpNot     bool
```
Comment (brett19) [inline transform.go]: This is from the old implementation of NotExpr, which is already completed

File: transform.go [hunk 5]
```diff
@@ -79,19 +80,23 @@ func opTypeToString(value OpType) string {
...
 	Rhs       interface{}
 }
 
 func (op OpNode) String() string {
 	var out string
-	out += fmt.Sprintf("[%d] %s", op.BucketIdx, opTypeToString(op.Op))
+	var opNotStr string
```
Comment (brett19) [inline transform.go]: This is from the old implementation of NotExpr, which is already completed

File: transform.go [hunk 6]
```diff
@@ -431,6 +436,15 @@ func (t *Transformer) makeRhsParam(expr Expression) interface{} {
 	}
 }
 
+func (t *Transformer) previousNotNode() bool {
```
Comment (brett19) [inline transform.go]: This is from the old implementation of NotExpr, which is already completed

File: transform.go [hunk 7]
```diff
@@ -513,6 +530,8 @@ func (t *Transformer) transformOne(expr Expression) *ExecNode {
 		return t.transformOr(expr)
 	case AndExpr:
 		return t.transformAnd(expr)
+	case NotExpr:
```
Comment (brett19) [inline transform.go]: This is a duplicate case from the merge between the new NotExpr implementation and the one originally in this commit.

File: matcher.go (hunk 1 above)
Comment (nelio2k) [reply matcher.go]: Removed.

File: expression_json.go (hunk 2 above)
Comment (nelio2k) [reply expression_json.go]: Removed

File: transform.go (hunk 4 above)
Comment (nelio2k) [reply transform.go]: Removed

File: transform.go (hunk 3 above)
Comment (nelio2k) [reply transform.go]: Agreed. I'm not sure why they showed up in the diff.

File: transform.go (hunk 5 above)
Comment (nelio2k) [reply transform.go]: Removed

File: transform.go (hunk 6 above)
Comment (nelio2k) [reply transform.go]: Removed

File: transform.go (hunk 7 above)
Comment (nelio2k) [reply transform.go]: Removed

**Pull request closed without merging**